### PR TITLE
Move Float16 functions away from float16.jl to places where other flo…

### PR DESCRIPTION
**Pull request by pkofod**
…at methods are defined.

As proposed in #17440 . 

Should the tests also be moved from `test/float16.jl`?

### CODE REVIEW

**Comment by pkofod** [comment]
Test failure on OSX Travis seems unrelated, but somewhat problematic. Known issue ? @tkelman 
```
WARNING: Error during initialization of module LinAlg:
ErrorException("could not load library "libopenblas"
dlopen(libopenblas.dylib, 1): image not found")
	From worker 7:	     * cmdlineargs           in 141.13 seconds, maxrss  560.18 MB
Exception running test spawn :
On worker 6:
LoadError: InterruptException:
 in include_string at ./loading.jl:434
 in jlcall_include_string_37292 at /private/tmp/julia/lib/julia/sys.dylib:?
 in include_from_node1 at ./loading.jl:484
 in jlcall_include_from_node1_20181 at /private/tmp/julia/lib/julia/sys.dylib:?
 in macro expansion at ./util.jl:226 [inlined]
 in runtests at /private/tmp/julia/share/julia/test/testdefs.jl:7
 in #629 at ./multi.jl:1421
 in run_work_thunk at ./multi.jl:1001
 in macro expansion at ./multi.jl:1421 [inlined]
 in #628 at ./event.jl:68
while loading /private/tmp/julia/share/julia/test/spawn.jl, in expression starting on line 75
ERROR: LoadError: Some tests exited with errors.
 in (::##11#18)() at /private/tmp/julia/share/julia/test/runtests.jl:72
 in cd(::##11#18, ::String) at ./file.jl:59
 in include_from_node1(::String) at ./loading.jl:481
 in include_from_node1(::String) at /private/tmp/julia/lib/julia/sys.dylib:?
 in process_options(::Base.JLOptions) at ./client.jl:262
 in _start() at ./client.jl:318
 in _start() at /private/tmp/julia/lib/julia/sys.dylib:?
while loading /private/tmp/julia/share/julia/test/runtests.jl, in expression starting on line 26
```

File: base/bool.jl [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ## boolean conversions ##
 
 convert(::Type{Bool}, x::Bool) = x
+convert(::Type{Bool}, x::Float16) = x==0 ? false : x==1 ? true : throw(InexactError())
```
Comment (vchuravy) [inline base/bool.jl]: `Float16 <: Real` so this definition is not needed?

File: base/float.jl [hunk 2]
```diff
@@ -349,16 +484,22 @@ end
...
+isnan(x::Float16)    = reinterpret(UInt16,x)&0x7fff  > 0x7c00
 isnan(x::Real) = false
 
 isfinite(x::AbstractFloat) = x - x == 0
+isfinite(x::Float16) = reinterpret(UInt16,x)&0x7c00 != 0x7c00
 isfinite(x::Real) = decompose(x)[3] != 0
 isfinite(x::Integer) = true
 
+
+isinf(x::Float16) = reinterpret(UInt16,x)&0x7fff == 0x7c00
```
Comment (vchuravy) [inline base/float.jl]: The definition below should also be correct for `Float16`.

Comment (pkofod) [inline base/float.jl]: That is correct! With isnan and isfinite defined, this definition is unnecessary.

File: base/bool.jl (hunk 1 above)
Comment (pkofod) [inline base/bool.jl]: That is correct, it is indeed superfluous. Thanks for the comments.

**Comment by pkofod** [comment]
Thanks @vchuravy ! I've removed the superfluous method definitions.

File: base/bool.jl (hunk 1 above)
Comment (pkofod) [inline base/bool.jl]: Actually, it was not! There was an ambiguity error from convert after removing this, so it needs to stay, as it is more specific than the other two ( can't remember where they were defined, but you can rebuild without this line and check).

File: base/float.jl [hunk 3]
```diff
@@ -129,13 +131,110 @@ function convert(::Type{Float32}, x::Int128)
     reinterpret(Float32, s | d + y)
 end
 
+function convert(::Type{Float16}, val::Float32)
+    f = reinterpret(UInt32, val)
+    i = (f >> 23) & 0x1ff + 1
+    sh = shifttable[i]
+    f &= 0x007fffff
+    h::UInt16 = basetable[i] + (f >> sh)
```
Comment (ararslan) [inline base/float.jl]: Could the `basetable` and `shifttable` lookups benefit from `@inbounds`?

File: base/float.jl [hunk 4]
```diff
@@ -129,13 +131,110 @@ function convert(::Type{Float32}, x::Int128)
...
+    # round
+    # NOTE: we maybe should ignore NaNs here, but the payload is
+    # getting truncated anyway so "rounding" it might not matter
+    nextbit = (f >> (sh-1)) & 1
+    if nextbit != 0
+        if h&1 == 1 ||  # round halfway to even
+            (f & ((1<<(sh-1))-1)) != 0  # check lower bits
+            h += 1
```
Comment (ararslan) [inline base/float.jl]: The comments are helpful but IMHO splitting the condition like that makes it harder to read. Could maybe do something like
```julia
# Round halfway to even, or check lower bits
if h & 1 == 1 || (f & ((1 << (sh - 1)) - 1)) != 0
    h += 1
end
```
?

Comment (pkofod) [inline base/float.jl]: Yes, I am merely re-organizing, but if people have useful suggestions, I'm fine with handling them in this PR if more senior people think that's fine. If not, then maybe it's more "next PR" material :)

**Comment by pkofod** [comment]
I pushed the changes to the control flow condition as proposed in the comments.

**Comment by tkelman** [comment]
appveyor should be restarted once https://github.com/JuliaLang/julia/pull/18397 is merged and we get new nightlies built from that

**Comment by pkofod** [comment]
AppVeyor worked, great.

**Comment by pkofod** [comment]
Rebased due to changes in `base/rational.jl`.

**Review by tkelman — APPROVED** [review]

File: base/bool.jl (hunk 1 above)
Comment (vchuravy) [inline base/bool.jl]: For the record:

```
  MethodError: convert(::Type{Bool}, ::Float16) is ambiguous. Candidates:
    convert{T<:Integer}(::Type{T}, x::Float16) at float.jl:66
    convert(::Type{Bool}, x::Real) at bool.jl:6
```

File: base/float.jl (hunk 3 above)
Comment (vchuravy) [inline base/float.jl]: Probably but we can do this later or do #17344

File: base/float.jl [hunk 5]
```diff
@@ -204,23 +303,31 @@ trunc(::Type{Unsigned}, x::Float32) = trunc(UInt,x)
 trunc(::Type{Unsigned}, x::Float64) = trunc(UInt,x)
 trunc(::Type{Integer}, x::Float32) = trunc(Int,x)
 trunc(::Type{Integer}, x::Float64) = trunc(Int,x)
+trunc{T<:Integer}(::Type{T}, x::Float16) = trunc(T, Float32(x))
 
 # fallbacks
 floor{T<:Integer}(::Type{T}, x::AbstractFloat) = trunc(T,floor(x))
+floor{T<:Integer}(::Type{T}, x::Float16) = floor(T, Float32(x))
```
Comment (vchuravy) [inline base/float.jl]: We could remove most of these definitions if we re-enable support for `Float16` intrinsics (which was disabled back in 2013 71ae4931e)

**Review by vchuravy — APPROVED** [review]
LGTM. There is a bit of cleanup left but we can do this later.

**Comment by vchuravy** [comment]
@StefanKarpinski I would like to merge this soon, because I am planning to work on lowering Float16 support back to intrinsics so that we have better support for them on the GPU.

**Comment by kshyatt** [comment]
https://gist.github.com/kshyatt/2df3743de28cf7f52e05881e1bc074c6 Gist of bad build

**Comment by vchuravy** [comment]
Do we still want to backport this?

**Comment by tkelman** [comment]
does it fix anything?

**Comment by vchuravy** [comment]
No it is a code reorganisation.

**Comment by StefanKarpinski** [comment]
How about we let it sit for a week and if there's no issue, then backport it to 0.5?

**Comment by StefanKarpinski** [comment]
Just for the sake of keeping the diff between master and 0.5 as small as we can.

**Comment by vchuravy** [comment]
We should make sure not to accidentally backport  #17297

**Comment by tkelman** [comment]
We're past the point where 0.5 and 0.6 have real code changes. I can usually handle conflicts if they happen in backporting something nearby, and when I can't I'll ask the author for help resolving them (or backport this at the time, if that's easiest). Making the diff smaller is not a very good reason to backport a code reorganization, there's more risk of accidentally changing something that shouldn't be than benefit.

**Comment by StefanKarpinski** [comment]
You're the backport boss, @tkelman, so whatever you prefer.

**Pull request merged**